src/lsm/groove.zig

```diff
@@ -703,36 +720,31 @@ pub fn GrooveType(
             /// I/O depth of the Grid.
             workers: [Grid.read_iops_max]PrefetchWorker = undefined,
             /// The number of workers that are currently running in parallel.
-            workers_busy: u32 = 0,
+            workers_pending: u32 = 0,
 
             fn start_workers(context: *PrefetchContext) void {
-                assert(context.workers_busy == 0);
+                assert(context.workers_pending == 0);
 
-                // Track an extra "worker" that will finish after the loop.
-                //
-                // This prevents `context.finish()` from being called within the loop body when
-                // every worker finishes synchronously. `context.finish()` calls the user-provided
-                // callback which may re-use the memory of this `PrefetchContext`. However, we
-                // rely on `context` being well-defined for the loop condition.
-                context.workers_busy += 1;
-
-                for (&context.workers) |*worker| {
+                for (&context.workers, 0..) |*worker, i| {
                     worker.* = .{ .context = context };
-                    context.workers_busy += 1;
+                    context.workers_pending += 1;
                     worker.lookup_start_next();
+
+                    // If the worker finished synchronously (e.g `workers_pending` decreased),
+                    // we don't need to start new ones.
+                    if (context.workers_pending == i) break;
                 }
 
```

Comment:
I am not sure if it is possible for no workers to start ...

It is possible for the first worker to start, complete synchronously calling the callback, and then we break the for loop and exit.

Reply:
Right, and in that case how do we prevent `prefetch()` from calling its callback synchronously?

Reply:
Good point,

Reply:
We call that synchronously on `main`, there's no `next_tick` involved:

https://github.com/tigerbeetle/tigerbeetle/blob/39fc9ace579c552fc89a544a732ef2105c40728f/src/lsm/groove.zig#L725-L726

Reply:
Ah, good catch. I wonder if that was intentional or not. :thinking:

Reply:
Well, I assumed that it was intentional, sometimes it's fine because we are already running from a callback.
But I checked and it's not the case, I guess we need to fix that one!